File: apps/desktop/src/git/git.ts
/**
 * @title Git data
 * @notice Typed wrappers over the backend git commands behind the history and
 * diff views.
 * @dev Read-only — nothing here writes to the repository. Commands shell out to
 * the `git` binary, so a missing or broken install surfaces as a plain error
 * message in the view instead of a crash.
 */
import { invoke, isTauri } from "@tauri-apps/api/core";

/** One row of the history list. */
export interface GitCommit {
  hash: string;
  short: string;
  subject: string;
  authorName: string;
  authorEmail: string;
  /** ISO-8601 author date. */
  authoredAt: string;
  /** Decorations such as `HEAD -> main` or `tag: v0.1.0`. */
  refs: string[];
  /** Parent count; more than one means a merge commit. */
  parents: number;
}

/** Repository facts for the history header. */
export interface GitRepoInfo {
  branch: string;
  remoteUrl: string | null;
}

/** One local branch, for the history header's branch picker. */
export interface GitBranch {
  name: string;
  /** Whether this is the checked-out branch. */
  current: boolean;
  /** Subject of the branch tip. */
  subject: string;
  /** ISO-8601 committer date of the branch tip. */
  committedAt: string;
  /** Upstream name such as `origin/main`, when configured. */
  upstream: string | null;
}

/** One entry of the working tree status. */
export interface GitChange {
  /** Path relative to the repository root. */
  path: string;
  /** Index (staged) status letter, e.g. `M`, `A`, `?`. */
  indexStatus: string;
  /** Work tree status letter. */
  worktreeStatus: string;
  /** Whether the change is staged and would go into the next commit. */
  staged: boolean;
}

/** Header plus patch of a single commit. */
export interface GitCommitDetail {
  hash: string;
  short: string;
  subject: string;
  body: string;
  authorName: string;
  authorEmail: string;
  authoredAt: string;
  /** Unified diff text; empty for merges without a recorded diff. */
  diff: string;
}

const NOT_DESKTOP = "Git history is only available in the desktop app.";

/**
 * @notice Lists commits of the repository containing `root`, newest first.
 * @param root Any directory inside the repository.
 * @param limit Maximum number of commits (the backend clamps to 1..=500).
 * @return Commit rows, or an empty list outside the desktop app.
 */
export async function gitLog(root: string, limit = 100): Promise<GitCommit[]> {
  if (!isTauri()) {
    return [];
  }
  return invoke<GitCommit[]>("git_log", { root, limit });
}

/**
 * @notice Reads the branch and origin URL of the repository containing `root`.
 * @param root Any directory inside the repository.
 * @return Repository info; rejects when `root` is not a work tree.
 */
export async function gitRepoInfo(root: string): Promise<GitRepoInfo> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  return invoke<GitRepoInfo>("git_repo_info", { root });
}

/**
 * @notice Lists the repository's local branches, newest tip first.
 * @param root Any directory inside the repository.
 * @return Branch rows, or an empty list outside the desktop app.
 */
export async function gitBranches(root: string): Promise<GitBranch[]> {
  if (!isTauri()) {
    return [];
  }
  return invoke<GitBranch[]>("git_branches", { root });
}

/**
 * @notice Checks out another local branch.
 * @dev The only command here that writes to the repository; git's own message
 * (for example a dirty working tree) is what the caller shows on failure.
 * @param root Any directory inside the repository.
 * @param branch Local branch name to check out.
 */
export async function gitCheckout(root: string, branch: string): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_checkout", { root, branch });
}

/**
 * @notice Reads the working tree status.
 * @param root Any directory inside the repository.
 * @return One row per changed path.
 */
export async function gitStatus(root: string): Promise<GitChange[]> {
  if (!isTauri()) {
    return [];
  }
  return invoke<GitChange[]>("git_status", { root });
}

/**
 * @notice Lists commits that are not on the upstream branch yet.
 * @dev Rejects when no upstream is configured; the caller decides what to show.
 * @param root Any directory inside the repository.
 * @return Unpushed commits, newest first.
 */
export async function gitUnpushed(root: string): Promise<GitCommit[]> {
  if (!isTauri()) {
    return [];
  }
  return invoke<GitCommit[]>("git_unpushed", { root });
}

/**
 * @notice Stages one path.
 * @param root Any directory inside the repository.
 * @param path Path relative to the repository root.
 */
export async function gitStage(root: string, path: string): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_stage", { root, path });
}

/**
 * @notice Unstages one path, leaving the working tree as it is.
 * @param root Any directory inside the repository.
 * @param path Path relative to the repository root.
 */
export async function gitUnstage(root: string, path: string): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_unstage", { root, path });
}

/**
 * @notice Commits what is staged, with the modifiers the panel offers.
 * @param root Any directory inside the repository.
 * @param message Commit message; may be empty when amending.
 * @param options Amend, sign-off, and hook-skipping flags.
 */
export async function gitCommit(
  root: string,
  message: string,
  options: { amend?: boolean; signoff?: boolean; noVerify?: boolean } = {},
): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_commit", {
    root,
    message,
    amend: options.amend === true,
    signoff: options.signoff === true,
    noVerify: options.noVerify === true,
  });
}

/**
 * @notice Lists the configured remotes.
 * @param root Any directory inside the repository.
 * @return Remote names.
 */
export async function gitRemotes(root: string): Promise<string[]> {
  if (!isTauri()) {
    return [];
  }
  return invoke<string[]>("git_remotes", { root });
}

/**
 * @notice Fetches from a remote, or from the default set when none is named.
 * @param root Any directory inside the repository.
 * @param remote Remote to fetch from.
 */
export async function gitFetch(root: string, remote?: string): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_fetch", { root, remote: remote ?? null });
}

/**
 * @notice Pulls the current branch from its upstream.
 * @param root Any directory inside the repository.
 * @param rebase Rebase local commits instead of merging.
 */
export async function gitPull(root: string, rebase: boolean): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_pull", { root, rebase });
}

/**
 * @notice Pushes one branch to a named remote.
 * @param root Any directory inside the repository.
 * @param remote Remote to push to.
 * @param branch Branch to push.
 */
export async function gitPushTo(
  root: string,
  remote: string,
  branch: string,
): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_push_to", { root, remote, branch });
}

/**
 * @notice Force-pushes the current branch with `--force-with-lease`.
 * @param root Any directory inside the repository.
 */
export async function gitForcePush(root: string): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_force_push", { root });
}

/**
 * @notice Pushes the checked-out branch to its upstream.
 * @dev Network call through git; failures carry git's own message.
 * @param root Any directory inside the repository.
 */
export async function gitPush(root: string): Promise<void> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  await invoke("git_push", { root });
}

/**
 * @notice Reads one commit: header plus unified diff.
 * @param root Any directory inside the repository.
 * @param hash Commit hash to read.
 * @return The commit detail.
 */
export async function gitCommitDetail(root: string, hash: string): Promise<GitCommitDetail> {
  if (!isTauri()) {
    throw new Error(NOT_DESKTOP);
  }
  return invoke<GitCommitDetail>("git_commit_detail", { root, hash });
}

File: apps/desktop/src-tauri/src/git.rs
//! @title Git history
//! @notice Read-only git queries behind the history and diff views.
//! @dev Shells out to the `git` binary instead of linking a library: the
//! commands are read-only, the output formats are stable, and it keeps the
//! binary small. Every call runs with `-C <root>`, so the process working
//! directory never changes and commands stay parallel-safe.

use std::process::Command;

use serde::Serialize;

/// One row of the history list.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommit {
    /// Full commit hash.
    pub hash: String,
    /// Abbreviated hash.
    pub short: String,
    /// First line of the message.
    pub subject: String,
    pub author_name: String,
    pub author_email: String,
    /// Author date in ISO-8601 (strict), as git reports it.
    pub authored_at: String,
    /// Decorations such as `HEAD -> main` or `tag: v0.1.0`, trimmed.
    pub refs: Vec<String>,
    /// Parent count; more than one means a merge commit.
    pub parents: usize,
}

/// Header plus patch of a single commit.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitDetail {
    pub hash: String,
    pub short: String,
    pub subject: String,
    /// Everything after the subject line, trimmed.
    pub body: String,
    pub author_name: String,
    pub author_email: String,
    pub authored_at: String,
    /// Unified diff text, without the commit header or diffstat.
    pub diff: String,
}

/// Repository facts for the history header.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitRepoInfo {
    /// Current branch, or `HEAD` when the branch is unborn or detached.
    pub branch: String,
    /// `origin` URL when one is configured.
    pub remote_url: Option<String>,
}

/// One local branch, for the history header's branch picker.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBranch {
    pub name: String,
    /// Whether this is the checked-out branch.
    pub current: bool,
    /// Subject of the branch tip.
    pub subject: String,
    /// Committer date of the branch tip, ISO-8601 strict.
    pub committed_at: String,
    /// Upstream name such as `origin/main`, when configured.
    pub upstream: Option<String>,
}

/// One entry of the working tree status.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitChange {
    /// Path relative to the repository root.
    pub path: String,
    /// Index (staged) status letter, e.g. `M`, `A`, `?`.
    pub index_status: String,
    /// Work tree status letter.
    pub worktree_status: String,
    /// Whether the change is staged and would go into the next commit.
    pub staged: bool,
}

/// Field separator: cannot occur in a commit subject or author name.
const FIELD: char = '\u{1f}';

/// One line per commit, fields split by `FIELD`, subject last so an empty
/// subject still yields the expected field count.
const LOG_FORMAT: &str = "--pretty=format:%H%x1f%h%x1f%an%x1f%ae%x1f%aI%x1f%P%x1f%D%x1f%s";

/// Largest patch handed to the UI, keeping a runaway commit from freezing it.
const MAX_DIFF_BYTES: usize = 400_000;

/// Runs git in `root` and returns stdout, mapping failures to the stderr text.
fn run(root: &str, args: &[&str]) -> Result<String, String> {
    let output = Command::new("git")
        .arg("-C")
        .arg(root)
        .args(args)
        .output()
        .map_err(|cause| format!("Could not run git: {cause}"))?;
    if !output.status.success() {
        let message = String::from_utf8_lossy(&output.stderr).trim().to_string();
        return Err(if message.is_empty() {
            "git exited with an error".to_string()
        } else {
            message
        });
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

/// Rejects a remote or branch name git would read as a flag.
fn checked(name: &str) -> Result<(), String> {
    if name.is_empty() || name.starts_with('-') {
        return Err("Invalid name".to_string());
    }
    Ok(())
}

/// Everything from the first `diff --git` marker, dropping the commit header.
fn patch_only(output: &str) -> String {
    match output.find("diff --git ") {
        Some(index) => output[index..].to_string(),
        None => String::new(),
    }
}

/// Caps a patch and notes the cut so the reader knows the diff is incomplete.
fn capped(patch: String) -> String {
    if patch.len() <= MAX_DIFF_BYTES {
        return patch;
    }
    let cut = patch[..MAX_DIFF_BYTES]
        .rfind('\n')
        .unwrap_or(MAX_DIFF_BYTES);
    format!("{}\n… diff truncated by ReNonce\n", &patch[..cut])
}

/// Parses `LOG_FORMAT` output into commit rows.
fn parse_log(output: &str) -> Vec<GitCommit> {
    output
        .lines()
        .filter(|line| !line.is_empty())
        .filter_map(|line| {
            let mut fields = line.split(FIELD);
            let hash = fields.next()?.to_string();
            let short = fields.next()?.to_string();
            let author_name = fields.next()?.to_string();
            let author_email = fields.next()?.to_string();
            let authored_at = fields.next()?.to_string();
            let parents = fields.next()?.split_whitespace().count();
            let refs = fields
                .next()?
                .split(',')
                .map(|value| value.trim().to_string())
                .filter(|value| !value.is_empty())
                .collect();
            let subject = fields.next().unwrap_or_default().to_string();
            Some(GitCommit {
                hash,
                short,
                subject,
                author_name,
                author_email,
                authored_at,
                refs,
                parents,
            })
        })
        .collect()
}

/// @notice Lists commits, newest first.
/// @param root Any directory inside the repository.
/// @param limit Maximum number of commits (clamped to 1..=500).
/// @return Commit rows.
#[tauri::command]
pub fn git_log(root: String, limit: usize) -> Result<Vec<GitCommit>, String> {
    let max_count = format!("--max-count={}", limit.clamp(1, 500));
    let output = run(&root, &["log", &max_count, LOG_FORMAT, "--no-color"])?;
    Ok(parse_log(&output))
}

/// @notice Lists commits that are not on the upstream branch yet.
/// @dev Errors when no upstream is configured; git's message is shown as is.
/// @param root Any directory inside the repository.
/// @return Unpushed commits, newest first.
#[tauri::command]
pub fn git_unpushed(root: String) -> Result<Vec<GitCommit>, String> {
    let output = run(
        &root,
        &["log", "@{upstream}..HEAD", LOG_FORMAT, "--no-color"],
    )?;
    Ok(parse_log(&output))
}

/// @notice Reads the working tree status.
/// @dev Porcelain v1 with NUL separators: paths keep their spaces, and a rename
/// or copy carries its source path as the next record, which is dropped here.
/// @param root Any directory inside the repository.
/// @return One row per changed path.
#[tauri::command]
pub fn git_status(root: String) -> Result<Vec<GitChange>, String> {
    let output = run(
        &root,
        &["status", "--porcelain=v1", "-z", "--untracked-files=all"],
    )?;
    let mut changes = Vec::new();
    let mut records = output.split('\0');
    while let Some(record) = records.next() {
        if record.len() < 4 {
            continue;
        }
        let index_status = record[0..1].to_string();
        let worktree_status = record[1..2].to_string();
        if index_status == "R" || index_status == "C" {
            records.next();
        }
        changes.push(GitChange {
            staged: index_status != " " && index_status != "?",
            path: record[3..].to_string(),
            index_status,
            worktree_status,
        });
    }
    Ok(changes)
}

/// @notice Stages one path.
/// @param root Any directory inside the repository.
/// @param path Path relative to the repository root.
#[tauri::command]
pub fn git_stage(root: String, path: String) -> Result<(), String> {
    run(&root, &["add", "--", &path])?;
    Ok(())
}

/// @notice Unstages one path, leaving the working tree as it is.
/// @param root Any directory inside the repository.
/// @param path Path relative to the repository root.
#[tauri::command]
pub fn git_unstage(root: String, path: String) -> Result<(), String> {
    run(&root, &["restore", "--staged", "--", &path])?;
    Ok(())
}

/// @notice Commits what is staged, with the modifiers the panel offers.
/// @dev `amend` folds the staged changes into the previous commit; an empty
/// message then keeps that commit's message. `signoff` adds a trailer, and
/// `no_verify` skips the repository's hooks.
/// @param root Any directory inside the repository.
/// @param message Commit message; required unless amending.
/// @param amend Amend the last commit instead of adding a new one.
/// @param signoff Append a `Signed-off-by` trailer.
/// @param no_verify Bypass pre-commit and commit-msg hooks.
#[tauri::command]
pub fn git_commit(
    root: String,
    message: String,
    amend: bool,
    signoff: bool,
    no_verify: bool,
) -> Result<(), String> {
    let trimmed = message.trim();
    if trimmed.is_empty() && !amend {
        return Err("A commit message is required".to_string());
    }
    let mut args: Vec<String> = vec!["commit".to_string(), "--quiet".to_string()];
    if amend {
        args.push("--amend".to_string());
    }
    if signoff {
        args.push("--signoff".to_string());
    }
    if no_verify {
        args.push("--no-verify".to_string());
    }
    if trimmed.is_empty() {
        args.push("--no-edit".to_string());
    } else {
        args.push("-m".to_string());
        args.push(trimmed.to_string());
    }
    let refs: Vec<&str> = args.iter().map(String::as_str).collect();
    run(&root, &refs)?;
    Ok(())
}

/// @notice Lists the configured remotes.
/// @param root Any directory inside the repository.
/// @return Remote names, e.g. `["origin"]`.
#[tauri::command]
pub fn git_remotes(root: String) -> Result<Vec<String>, String> {
    let output = run(&root, &["remote"])?;
    Ok(output
        .lines()
        .map(|line| line.trim().to_string())
        .filter(|line| !line.is_empty())
        .collect())
}

/// @notice Fetches from a remote, or from every remote when none is named.
/// @param root Any directory inside the repository.
/// @param remote Remote to fetch from; `None` fetches the default set.
#[tauri::command]
pub fn git_fetch(root: String, remote: Option<String>) -> Result<(), String> {
    match remote {
        Some(name) => {
            checked(&name)?;
            run(&root, &["fetch", &name])?;
        }
        None => {
            run(&root, &["fetch"])?;
        }
    }
    Ok(())
}

/// @notice Pulls the current branch from its upstream.
/// @param root Any directory inside the repository.
/// @param rebase Rebase local commits instead of creating a merge commit.
#[tauri::command]
pub fn git_pull(root: String, rebase: bool) -> Result<(), String> {
    if rebase {
        run(&root, &["pull", "--rebase"])?;
    } else {
        run(&root, &["pull"])?;
    }
    Ok(())
}

/// @notice Pushes one branch to a named remote.
/// @param root Any directory inside the repository.
/// @param remote Remote to push to.
/// @param branch Branch to push.
#[tauri::command]
pub fn git_push_to(root: String, remote: String, branch: String) -> Result<(), String> {
    checked(&remote)?;
    checked(&branch)?;
    run(&root, &["push", &remote, &branch])?;
    Ok(())
}

/// @notice Force-pushes the current branch, refusing to clobber unseen work.
/// @dev Uses `--force-with-lease`, never a bare `--force`: git rejects the push
/// when the remote moved since the last fetch.
/// @param root Any directory inside the repository.
#[tauri::command]
pub fn git_force_push(root: String) -> Result<(), String> {
    run(&root, &["push", "--force-with-lease"])?;
    Ok(())
}

/// @notice Pushes the checked-out branch to its upstream.
/// @dev Reaches the network through git; git's own failure text is returned.
/// @param root Any directory inside the repository.
#[tauri::command]
pub fn git_push(root: String) -> Result<(), String> {
    run(&root, &["push"])?;
    Ok(())
}

/// @notice Describes the repository at `root`.
/// @param root Any directory inside the repository.
/// @return Branch and origin URL; errors when `root` is not a work tree.
#[tauri::command]
pub fn git_repo_info(root: String) -> Result<GitRepoInfo, String> {
    run(&root, &["rev-parse", "--is-inside-work-tree"])?;
    let branch = run(&root, &["rev-parse", "--abbrev-ref", "HEAD"])
        .or_else(|_| run(&root, &["symbolic-ref", "--short", "HEAD"]))
        .unwrap_or_else(|_| "HEAD".to_string())
        .trim()
        .to_string();
    let remote_url = run(&root, &["config", "--get", "remote.origin.url"])
        .ok()
        .map(|value| value.trim().to_string())
        .filter(|value| !value.is_empty());
    Ok(GitRepoInfo { branch, remote_url })
}

/// @notice Lists local branches, newest tip first.
/// @param root Any directory inside the repository.
/// @return Branch rows; the checked-out branch is flagged.
#[tauri::command]
pub fn git_branches(root: String) -> Result<Vec<GitBranch>, String> {
    let format = format!(
        "--format=%(refname:short){0}%(HEAD){0}%(contents:subject){0}%(committerdate:iso-strict){0}%(upstream:short)",
        FIELD
    );
    let output = run(
        &root,
        &[
            "for-each-ref",
            "--sort=-committerdate",
            &format,
            "refs/heads",
        ],
    )?;
    let branches = output
        .lines()
        .filter(|line| !line.is_empty())
        .filter_map(|line| {
            let mut fields = line.split(FIELD);
            let name = fields.next()?.to_string();
            let current = fields.next()?.trim() == "*";
            let subject = fields.next()?.to_string();
            let committed_at = fields.next()?.to_string();
            let upstream = fields.next().unwrap_or_default().trim().to_string();
            Some(GitBranch {
                name,
                current,
                subject,
                committed_at,
                upstream: if upstream.is_empty() {
                    None
                } else {
                    Some(upstream)
                },
            })
        })
        .collect();
    Ok(branches)
}

/// @notice Switches the working tree to another local branch.
/// @dev Failures (dirty tree, unknown branch) come back as git's own message.
/// @param root Any directory inside the repository.
/// @param branch Local branch name to check out.
#[tauri::command]
pub fn git_checkout(root: String, branch: String) -> Result<(), String> {
    checked(&branch)?;
    run(&root, &["checkout", "--quiet", &branch])?;
    Ok(())
}

/// @notice Reads one commit: its header and its unified diff.
/// @param root Any directory inside the repository.
/// @param hash Commit hash to show.
/// @return Commit header plus patch; the patch is empty for merges.
#[tauri::command]
pub fn git_commit_detail(root: String, hash: String) -> Result<GitCommitDetail, String> {
    let meta_format = format!("--pretty=format:%H{0}%h{0}%an{0}%ae{0}%aI{0}%s{0}%b", FIELD);
    let meta = run(
        &root,
        &["show", "--no-patch", "--no-color", &meta_format, &hash],
    )?;
    let mut fields = meta.split(FIELD);
    let full = fields.next().unwrap_or_default().to_string();
    let short = fields.next().unwrap_or_default().to_string();
    let author_name = fields.next().unwrap_or_default().to_string();
    let author_email = fields.next().unwrap_or_default().to_string();
    let authored_at = fields.next().unwrap_or_default().to_string();
    let subject = fields.next().unwrap_or_default().to_string();
    let body = fields.next().unwrap_or_default().trim().to_string();

    let patch = run(
        &root,
        &["show", "--patch", "--no-color", "--pretty=format:", &hash],
    )?;

    Ok(GitCommitDetail {
        hash: full,
        short,
        subject,
        body,
        author_name,
        author_email,
        authored_at,
        diff: capped(patch_only(&patch)),
    })
}
